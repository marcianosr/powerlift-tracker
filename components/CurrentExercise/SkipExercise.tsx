import React, { FC } from "react";
import Button from "../Button";
import Card from "../Card";
import Title from "../Title";
import ChevronIcon from "../../public/icons/chevron.svg";
import styles from "./styles.module.scss";

type SkipExerciseProps = {
	exercise: string;
	onSkip: () => void;
};

const SkipExercise: FC<SkipExerciseProps> = ({ exercise, onSkip }) => {
	return (
		<>
			<Card>
				<Title tag="h3" variant="small" shade="light">
					{exercise}
				</Title>
				<span className={styles.subTitle}>
					It's skip "{exercise}" day!
				</span>
			</Card>
			<div className={styles.buttons}>
				<Button
					variant="large"
					onClick={onSkip}
					align="right"
					iconRight={<ChevronIcon width={15} height={15} />}
				>
					Go to next
				</Button>
			</div>
		</>
	);
};
export default SkipExercise;
