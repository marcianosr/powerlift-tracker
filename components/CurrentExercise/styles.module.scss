.container {
	gap: 0.5rem;
	display: flex;
	flex-direction: column;
	margin-top: 2rem;
}

.hashIcon {
	color: white;
	font-size: 1rem;
}

.subTitle {
	--textColor: var(--gray-30);

	text-transform: uppercase;
	font-size: 1rem;
	font-weight: 700;
	color: var(--textColor);
}

.weightContent {
	display: flex;
	gap: 0.75rem;
	margin-top: 0.5rem;
	align-items: center;
}

.actionsContainer {
	display: flex;
	gap: 0.25rem;
}

.buttons {
	display: flex;
	flex-direction: column;
	gap: 0.5rem;

	&:first-child {
		flex: 1;
	}
}

.weightText {
	font-size: 2.75rem;
	font-weight: 800;
	margin: 0;
	margin-left: 1rem;

	span {
		font-size: 1.5rem;
	}
}

.bar {
	width: 100%;
	height: 30px;
	background-color: var(--gray-30);
	// position: absolute;
	// transform: translate(0%, calc(100% + 30px));
}

.weightIndicatorContainer {
	// display: flex;
	display: grid;
	grid-template-columns: 2fr 1fr 1fr;
	align-items: center;
	// justify-self: end;
	justify-content: space-between;
}

.scrollablePlates {
	overflow-x: scroll;
	display: flex;
	gap: 1rem;
	font-size: 2rem;
	font-weight: 800;
	padding-left: 1rem;
	padding-right: 1rem;
}
