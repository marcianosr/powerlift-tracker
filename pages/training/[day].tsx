import styles from "@/pages/index.module.css";
import { GetServerSideProps, NextPage } from "next";
import Layout from "@/components/Layout";
import { useRouter } from "next/router";
import ExcersiseListItem from "@/components/ExcersiseListItem";
import { ExceriseInfoItems } from "@/components/ExcersiseListItem/types";
import {
	GoogleSpreadsheet,
	GoogleSpreadsheetWorksheet,
} from "google-spreadsheet";
import range from "lodash.range";

type TrainingPageProps = {
	data: any;
};

const TrainingPage: NextPage<TrainingPageProps> = ({ data }) => {
	// console.log(data);
	return <h1>Powerlift</h1>;
	const { query } = useRouter();

	const excersisesInfo: ExceriseInfoItems[] = data.map(
		([excersise, sets, reps, ...previous]: [
			string,
			string,
			string,
			string
		]) => ({
			excersise,
			sets,
			reps,
			weight: previous[previous.length - 3].split("x")[0],
			RPE: previous[previous.length - 2],
			previous: {
				weight: previous[previous.length - 4],
				RPE: previous[previous.length - 3],
			},
		})
	);

	return (
		<Layout>
			<h1>{query.day}</h1>
			<section className={styles.list}>
				{excersisesInfo.map((item, idx) => (
					<ExcersiseListItem item={item} idx={idx} key={idx} />
				))}
			</section>
		</Layout>
	);
};

type SlugKeys = "dag-a" | "dag-b" | "dag-c" | "dag-d";
type SlugValues = "A" | "B" | "C" | "D";
type SlugMappingTypes = {
	[key in SlugKeys]: SlugValues;
};

const DAY_KEYS_BY_ROUTE_MAPPING: SlugMappingTypes = {
	["dag-a"]: "A",
	["dag-b"]: "B",
	["dag-c"]: "C",
	["dag-d"]: "D",
};

const getBasicProgramInfo = async (sheet: GoogleSpreadsheetWorksheet) => {
	const startOffset = 5;
	const totalRowLength = (await sheet.getRows()).length;

	await sheet.loadCells(`A${startOffset}:D${totalRowLength}`); // loads a range of cells

	const rowIndexes = range(startOffset, totalRowLength);

	return rowIndexes.map((idx) => ({
		day: sheet.getCellByA1(`A${idx}`).value,
		excersise: sheet.getCellByA1(`B${idx}`).value,
		sets: sheet.getCellByA1(`C${idx}`).value,
		reps: sheet.getCellByA1(`D${idx}`).value,
	}));
};

export const getServerSideProps: GetServerSideProps = async ({ params }) => {
	const creds = require("../../powerlift-339515-4c4a82b74db1.json"); // the file saved above
	const doc = new GoogleSpreadsheet(
		"1CCNcbt-nEYZS9z_uowQK7SWmHCPAJGgQ9I8H92C-6Ck"
	);
	await doc.useServiceAccountAuth(creds);

	// or preferably, load that info from env vars / config instead of the file
	await doc.useServiceAccountAuth({
		client_email: process.env.GOOGLE_SERVICE_ACCOUNT_EMAIL || "",
		private_key:
			process.env.GOOGLE_PRIVATE_KEY?.replace(/\\n/g, "\n") || "",
	});

	await doc.loadInfo();
	const sheet = doc.sheetsByIndex[0];

	const excersiseData = await getBasicProgramInfo(sheet);

	const dayQuery = params?.day as SlugKeys;
	const slug = DAY_KEYS_BY_ROUTE_MAPPING[dayQuery];

	const program = excersiseData.filter((item) => item.day === slug);

	const data = {
		program,
	};
	return { props: { data } };
};

export default TrainingPage;
