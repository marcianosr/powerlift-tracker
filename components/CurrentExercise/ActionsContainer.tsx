import React, { FC, useState } from "react";
import Button from "../Button";
import WeightsIcon from "../../public/icons/weights.svg";
import DoneIcon from "../../public/icons/done.svg";
import RPEIcon from "../../public/icons/rpe.svg";
import ChevronIcon from "../../public/icons/chevron.svg";
import styles from "./styles.module.scss";
import PreviousResultsButton from "../PreviousResultsButton";
import Drawer from "../Drawer";
import ChangeLoadDrawerContent from "../ChangeLoadDrawerContent";
import { ExcelData } from "@/pages/training/[week]/[day]";
import { Bars, divideWeightForPlates } from "../WeightIndicator/utils";
import { useDataSheet } from "providers/SheetDataProvider";

type ActionsContainerProps = {
	markDone: () => void;
	currentExercise: ExcelData;
};

const ActionsContainer: FC<ActionsContainerProps> = ({
	markDone,
	currentExercise,
}) => {
	const [showDrawer, setShowDrawer] = useState({ type: "", open: false });

	const [loadedBar, setLoadedBar] = useState({
		plates: divideWeightForPlates(
			(currentExercise.result &&
				+currentExercise.result?.weight - Bars.Olympic) ||
				0
		),
		weight:
			(currentExercise.result && +currentExercise.result?.weight) || 0,
	});

	const { data, updateSheet } = useDataSheet();

	const getCurrentCellId = data.find(
		(item) => item?.result?.cellId === currentExercise.result?.cellId
	)?.result?.cellId;

	return (
		<section className={styles.actionsContainer}>
			<div className={styles.buttons}>
				<Button
					iconLeft={<WeightsIcon />}
					iconRight={<ChevronIcon width={15} height={15} />}
					onClick={() =>
						setShowDrawer({ type: "change-load", open: true })
					}
				>
					Change load
				</Button>
				<Button
					iconLeft={<span className={styles.hashIcon}>#</span>}
					iconRight={<ChevronIcon width={15} height={15} />}
				>
					Change reps
				</Button>
				<Button
					iconLeft={<RPEIcon />}
					iconRight={<ChevronIcon width={15} height={15} />}
				>
					Set RPE
				</Button>
			</div>
			<div className={styles.buttons}>
				<PreviousResultsButton />
				<Button
					variant="smallRound"
					onClick={markDone}
					align="right"
					testId="doneButton"
				>
					<DoneIcon width={33} height={28} />
				</Button>
			</div>
			{showDrawer.open && (
				<Drawer
					onClick={() =>
						updateSheet({
							cell: getCurrentCellId || "",
							value: {
								weight: loadedBar.weight,
							},
						})
					}
				>
					<ChangeLoadDrawerContent
						currentExercise={currentExercise}
						loadedBar={loadedBar}
						setLoadedBar={setLoadedBar}
					/>
				</Drawer>
			)}
		</section>
	);
};
export default ActionsContainer;
