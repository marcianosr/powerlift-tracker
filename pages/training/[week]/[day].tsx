import styles from "@/pages/index.module.css";
import { GetServerSideProps, NextPage } from "next";
import Layout from "@/components/Layout";
import { useRouter } from "next/router";
import { GoogleSpreadsheet } from "google-spreadsheet";
import { getBasicProgramInfo, loadColumnDataByLetter } from "../../../util";
import CurrentExercise from "@/components/CurrentExercise";
import Header from "@/components/Header";

export type ExcelData = {
	id: string;
	exercise: string;
	day: DayValues;
	reps: string;
	sets: number;
	type: "barbell" | "machine" | "dumbell" | "no-type";
	plan: Plan;
};

type Plan = {
	id: string;
	day: DayValues;
	weight: number | string;
	unit: string;
	reps: string[];
	RPE: string;
} | null;

type TrainingPageProps = {
	data: ExcelData[];
};

const TrainingPage: NextPage<TrainingPageProps> = ({ data }) => {
	const { query } = useRouter();

	const noTrainingPrescription = data.filter((item) => item.plan && []);

	return (
		<Layout>
			<Header pageTitle={query.day} />
			{noTrainingPrescription.length === 0 ? (
				<h1>No prescribed weights yet.</h1>
			) : (
				<CurrentExercise data={data} />
			)}
		</Layout>
	);
};

type SlugKeys = "dag-a" | "dag-b" | "dag-c" | "dag-d";
type DayValues = "A" | "B" | "C" | "D";
type SlugMappingTypes = {
	[key in SlugKeys]: DayValues;
};

const DAY_KEYS_BY_ROUTE_MAPPING: SlugMappingTypes = {
	["dag-a"]: "A",
	["dag-b"]: "B",
	["dag-c"]: "C",
	["dag-d"]: "D",
};

export const getServerSideProps: GetServerSideProps = async ({ params }) => {
	const creds = require("../../../powerlift-339515-4c4a82b74db1.json"); // the file saved above
	const doc = new GoogleSpreadsheet(
		"1CCNcbt-nEYZS9z_uowQK7SWmHCPAJGgQ9I8H92C-6Ck"
	);
	await doc.useServiceAccountAuth(creds);

	// or preferably, load that info from env vars / config instead of the file
	await doc.useServiceAccountAuth({
		client_email: process.env.GOOGLE_SERVICE_ACCOUNT_EMAIL || "",
		private_key:
			process.env.GOOGLE_PRIVATE_KEY?.replace(/\\n/g, "\n") || "",
	});

	await doc.loadInfo();
	const sheet = doc.sheetsByIndex[0]; // ? improve: get training by tab name
	const weekIndexFromQuery = params?.week || "";
	const excersiseData = (await getBasicProgramInfo(sheet)) as Exclude<
		ExcelData,
		"plan"
	>[];

	const programData = (await loadColumnDataByLetter(
		sheet,
		weekIndexFromQuery
	)) as (Plan | undefined)[];

	const dayQuery = params?.day as SlugKeys;
	const slug = DAY_KEYS_BY_ROUTE_MAPPING[dayQuery];

	const programExcersises = excersiseData.filter((item) => item.day === slug);
	const programResult = programData.filter((item) => item?.day === slug);

	const data = programExcersises
		.map((data) => {
			const match = programResult.find((t) => t?.id === data.id);

			if (!match) return { ...data, plan: null };

			return { ...data, plan: match };
		})
		.filter((item) => item);

	return { props: { data } };
};

export default TrainingPage;
