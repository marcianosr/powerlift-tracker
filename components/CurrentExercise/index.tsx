import React, { FC, useState } from "react";
import Card from "../Card";
import WeightLine from "../WeightLine";
import WeightIndicator from "../WeightIndicator";
import styles from "./styles.module.scss";
import RPEContainer from "../RPEContainer";
import Title from "../Title";
import { Bars, divideWeightForPlates } from "../WeightIndicator/utils";
import SkipExercise from "./SkipExercise";
import ActionsContainer from "./ActionsContainer";
import { ExcelData } from "@/pages/training/[week]/[day]";

type CurrentExerciseProps = {
	data: ExcelData[];
};

const CurrentExercise: FC<CurrentExerciseProps> = ({ data }) => {
	const [exercises, setExercises] = useState({
		total: 0,
		current: data[0],
		count: 1,
	});

	const [count, setCount] = useState(1);
	const [currentSet, setCurrentSet] = useState({
		total: exercises.current.sets,
		count: 1,
	});

	const isExerciseDone = currentSet.count === exercises.current.sets;
	const exerciseIsDone = () => {
		setCurrentSet({ ...currentSet, count: 1 });
		setCount(count + 1);

		return setExercises({
			...exercises,
			current: data[count],
			count: count,
		});
	};

	const exerciseHasMultipleSets =
		exercises.current.sets > 1 && currentSet.count < exercises.current.sets;

	const markDone = () => {
		if (isExerciseDone) return exerciseIsDone();

		if (exerciseHasMultipleSets) {
			return setCurrentSet({
				...currentSet,
				count: currentSet.count + 1,
			});
		}
	};

	const onSkip = () => {
		setCount(count + 1);

		return setExercises({
			...exercises,
			current: data[count],
			count: count,
		});
	};

	return (
		<section className={styles.container}>
			<Title shade="light" variant="x-small">
				Doing
			</Title>
			{!exercises.current.plan ? (
				<SkipExercise
					exercise={exercises.current.exercise}
					onSkip={onSkip}
				/>
			) : (
				<>
					<Card>
						<section>
							<Title tag="h3" variant="small" shade="light">
								{exercises.current.exercise}
							</Title>
							<div className={styles.weightContent}>
								<span className={styles.subTitle}>
									Set {currentSet.count}
								</span>
								{exercises.current.type === "barbell" && (
									<WeightIndicator
										weights={divideWeightForPlates(
											+exercises.current.plan.weight -
												Bars.Olympic
										)}
									/>
								)}
								<WeightLine
									weight={exercises.current.plan.weight}
									reps={exercises.current.reps}
								/>
								<RPEContainer
									RPE={exercises.current.plan.RPE || ""}
								/>
							</div>
						</section>
					</Card>
					<ActionsContainer markDone={markDone} />
				</>
			)}
		</section>
	);
};

export default CurrentExercise;
