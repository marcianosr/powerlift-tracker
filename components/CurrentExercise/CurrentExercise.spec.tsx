import { ExcelData } from "@/pages/training/[week]/[day]";
import { render } from "@testing-library/react";
import CurrentExercise from ".";

const mockData: ExcelData[] = [
	{
		id: "C20",
		day: "C",
		exercise: "1) Cluster Deadlift (1MIN RI)",
		sets: 6,
		reps: "1",
		type: "barbell",
		plan: {
			id: "C20",
			day: "C",
			weight: 252.5,
			unit: "kg",
			reps: ["1 / 265kg"],
			RPE: "9",
		},
	},
	{
		id: "C21",
		day: "C",
		exercise: "    Deadlift Volume ",
		sets: 2,
		reps: "4",
		type: "barbell",
		plan: {
			id: "C21",
			day: "C",
			weight: 227.5,
			unit: "kg",
			reps: ["1", " 1 (2"],
			RPE: "7.5",
		},
	},
	{
		id: "C22",
		day: "C",
		exercise: "2) Paused High Bar Squat ",
		sets: 3,
		reps: "5",
		type: "barbell",
		plan: {
			id: "C22",
			day: "C",
			weight: 165,
			unit: "kg",
			reps: ["5", " 5", " 5"],
			RPE: "8.5",
		},
	},
	{
		id: "C23",
		day: "C",
		exercise: "3) Hip Thrust machine ",
		sets: 3,
		reps: "10-12",
		type: "no-type",
		plan: null,
	},
	{
		id: "C24",
		day: "C",
		exercise: "4) Seated hamstring Curl ",
		sets: 3,
		reps: "15-20",
		type: "no-type",
		plan: null,
	},
	{
		id: "C25",
		day: "C",
		exercise: "5) Standing Calf Raise ",
		sets: 3,
		reps: "10-12",
		type: "machine",
		plan: {
			id: "C25",
			day: "C",
			weight: "Stand 13",
			unit: "unknown",
			reps: ["12", " 12", " 12"],
			RPE: "8.5",
		},
	},
	{
		id: "C26",
		day: "C",
		exercise: "6) Copenhagen Plank",
		sets: 3,
		reps: "ALAP",
		type: "no-type",
		plan: null,
	},
];

test("renders the card data", () => {
	const { asFragment } = render(<CurrentExercise data={mockData} />);

	expect(asFragment()).toMatchSnapshot();
});
