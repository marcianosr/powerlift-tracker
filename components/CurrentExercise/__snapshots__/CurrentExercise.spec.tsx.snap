// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the card data 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <span
      class="title light x-small"
    >
      Doing
    </span>
    <article
      class="card"
    >
      <section>
        <h3
          class="title light small"
        >
          1) Cluster Deadlift (1MIN RI)
        </h3>
        <div
          class="weightContent"
        >
          <span
            class="subTitle"
          >
            Set 1
          </span>
          <div
            class="weightIndicator"
            style="--heightMultiplier: 1;"
          >
            <div
              class="plate gray"
              data-testid="plate-1.25"
              style="--heightMultiplier: 1; width: 4px;"
            />
            <div
              class="plate yellow"
              data-testid="plate-15"
              style="--heightMultiplier: 1; width: 4px;"
            />
            <div
              class="plate red"
              data-testid="plate-25"
              style="--heightMultiplier: 1; width: 4px;"
            />
            <div
              class="plate red"
              data-testid="plate-25"
              style="--heightMultiplier: 1; width: 4px;"
            />
            <div
              class="plate red"
              data-testid="plate-25"
              style="--heightMultiplier: 1; width: 4px;"
            />
            <div
              class="plate red"
              data-testid="plate-25"
              style="--heightMultiplier: 1; width: 4px;"
            />
          </div>
          <div
            class="weightContainer"
          >
            <span
              class="weight"
            >
              252.5
            </span>
            <span
              class="kg"
            >
              kg
            </span>
             
            <span
              class="x"
            >
              x
            </span>
             
            <span
              class="reps"
              style="--textLength: 2.5rem;"
            >
              1
            </span>
          </div>
          <div
            class="container"
          >
            <span>
              @
            </span>
            <div
              class="RPEContainer redOrange"
            >
              9
            </div>
          </div>
        </div>
      </section>
    </article>
    <section
      class="actionsContainer"
    >
      <div
        class="buttons"
      >
        <button
          class="button undefined"
          role="button"
        >
          <test-file-stub />
          <span
            class="text"
          >
            Change load
          </span>
          <test-file-stub
            height="15"
            width="15"
          />
        </button>
        <button
          class="button undefined"
          role="button"
        >
          <span
            class="hashIcon"
          >
            #
          </span>
          <span
            class="text"
          >
            Change reps
          </span>
          <test-file-stub
            height="15"
            width="15"
          />
        </button>
        <button
          class="button undefined"
          role="button"
        >
          <test-file-stub />
          <span
            class="text"
          >
            Set RPE
          </span>
          <test-file-stub
            height="15"
            width="15"
          />
        </button>
      </div>
      <div
        class="buttons"
      >
        <button
          class="button large"
          role="button"
        >
          <span
            class="text"
          >
            <div
              class="previousContainer"
            >
              <span
                class="title darker small"
              >
                Previous
              </span>
              <span>
                <span>
                  185kg x 2
                </span>
                 
                <span>
                  @ 7.5
                </span>
                 
              </span>
            </div>
          </span>
          <test-file-stub
            height="15"
            width="15"
          />
        </button>
        <button
          class="button smallRound alignRight"
          data-testid="doneButton"
          role="button"
        >
          <span
            class="text"
          >
            <test-file-stub
              height="28"
              width="33"
            />
          </span>
        </button>
      </div>
    </section>
  </section>
</DocumentFragment>
`;
